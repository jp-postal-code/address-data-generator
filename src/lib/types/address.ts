export interface Address {
  postalCode: string;
  regionId: number;
  region: string;
  locality: string;
  address1: string;
  address2: string;
  regionKana: string;
  localityKana: string;
  address1Kana: string;
  address2Kana: string;
}
