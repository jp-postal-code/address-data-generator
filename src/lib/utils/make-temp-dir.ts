import { mkdtemp } from 'fs/promises';
import { tmpdir } from 'os';
import { join } from 'path';

export async function makeTempDir(): Promise<string> {
  return await mkdtemp(join(tmpdir(), 'adg-'));
}
