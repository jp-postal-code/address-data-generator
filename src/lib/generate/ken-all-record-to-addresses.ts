import { REGION_ID_MAP } from '../constants';
import { Address } from '../types/address';
import { KenAllRecord } from '../types/ken-all-record';

export function kenAllRecordToAddresses(record: KenAllRecord): Address[] {
  const regionId = REGION_ID_MAP.get(record[6]);

  if (regionId == null) {
    throw new Error(`"${record[6]}"は不明な都道府県です`);
  }

  const address = removeNonAddressString({
    postalCode: record[2],
    regionId,
    region: record[6],
    locality: record[7],
    address1: record[8],
    address2: '',
    regionKana: record[3].normalize('NFKC'),
    localityKana: record[4].normalize('NFKC'),
    address1Kana: record[5].normalize('NFKC'),
    address2Kana: ''.normalize('NFKC'),
  });

  return [address];
}

function removeNonAddressString(address: Address): Address {
  let { address1, address1Kana } = address;

  // 町域ではない文言の場合は空文字にする
  if (
    /以下に掲載がない場合$/.test(address1) ||
    /.+がくる場合/.test(address1) ||
    /.+一円/.test(address1)
  ) {
    address1 = '';
    address1Kana = '';
  }

  // 以下は住所として使える可能性があるが、一旦消しているもの
  // TODO 住所として使えるものは使う
  if (/[、（～]/.test(address1)) {
    address1 = '';
    address1Kana = '';
  }

  return { ...address, address1, address1Kana };
}
