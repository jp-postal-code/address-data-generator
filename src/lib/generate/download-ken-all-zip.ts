import { writeFile } from 'fs/promises';
import { KEN_ALL_ZIP_URL } from '../constants';
import { fetch } from '@/lib/utils/fetch';

/**
 * @returns ダウンロードファイルのパス
 */
export async function downloadKenAllZip(path: string): Promise<void> {
  const response = await fetch(KEN_ALL_ZIP_URL);

  if (response.statusCode !== 200) {
    throw new Error(
      `KEN_ALL.zip ファイルのダウンロードに失敗しました\nStatus code: ${response.statusCode}`
    );
  }

  const body = await response.body();

  await writeFile(path, body);
}
