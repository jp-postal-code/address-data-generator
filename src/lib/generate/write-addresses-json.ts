import { Address } from '../types/address';
import { PromisePool } from '@supercharge/promise-pool';
import { writeJsonFile } from '../utils/write-json-file';
import { dirname, join } from 'path';
import { mkdir } from 'fs/promises';

interface Params {
  postalCodeAddressesMap: Map<string, Address[]>;
  distPath: string;
  pretty: boolean;

  onOutput?(path: string): void;
}

export async function writeAddressesJson({
  postalCodeAddressesMap,
  distPath,
  pretty,
  onOutput,
}: Params): Promise<void> {
  await PromisePool.for([...postalCodeAddressesMap.entries()])
    .withConcurrency(50)
    .handleError((error) => {
      throw error;
    })
    .process(async ([postalCode, addresses]) => {
      const outputPath = join(
        distPath,
        postalCode.substring(0, 3),
        `${postalCode.substring(3)}.json`
      );

      await mkdir(dirname(outputPath), { recursive: true });

      await writeJsonFile(outputPath, addresses, pretty);

      onOutput?.(outputPath);
    });
}
