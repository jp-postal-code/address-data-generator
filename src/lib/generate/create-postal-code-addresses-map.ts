import { Address } from '../types/address';
import { KenAllRecord } from '../types/ken-all-record';
import { isUniqueFactory } from '../utils/is-unique';
import { kenAllRecordToAddresses } from './ken-all-record-to-addresses';

export async function createPostalCodeAddressesMap(
  kenAllRecordGenerator: AsyncGenerator<KenAllRecord>
): Promise<Map<string, Address[]>> {
  const result = new Map<string, Address[]>();
  const isUnique = isUniqueAddressFactory();

  for await (const record of kenAllRecordGenerator) {
    const addresses = kenAllRecordToAddresses(record);

    for (const address of addresses) {
      if (!isUnique(address)) {
        continue;
      }

      const mapValue = result.get(address.postalCode) ?? [];

      mapValue.push(address);

      result.set(address.postalCode, mapValue);
    }
  }

  return result;
}

function isUniqueAddressFactory() {
  const isUnique = isUniqueFactory<string>();

  return (address: Address) => isUnique(Object.values(address).join('_'));
}
