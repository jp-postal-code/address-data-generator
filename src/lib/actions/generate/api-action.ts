import { KEN_ALL_CSV_FILENAME } from '@/lib/constants';
import { createPostalCodeAddressesMap } from '@/lib/generate/create-postal-code-addresses-map';
import { downloadKenAllZip } from '@/lib/generate/download-ken-all-zip';
import { readKenAllRecord } from '@/lib/generate/read-ken-all-record';
import { writeAddressesJson } from '@/lib/generate/write-addresses-json';
import { Address } from '@/lib/types/address';
import { makeTempDir } from '@/lib/utils/make-temp-dir';
import { unzip } from '@/lib/utils/unzip';
import { rm } from 'fs/promises';
import { Listr } from 'listr2';
import { join } from 'path';
import { Action } from '../../types/action';

interface Options {
  pretty: boolean;
}

export const defaultOptions = {
  pretty: false,
};

export const apiAction: Action<[string], Options> = async (
  distPath,
  { pretty }
) => {
  const tempDir = await makeTempDir();
  const kenAllZipPath = join(tempDir, 'ken_all.zip');
  const kenAllCsvPath = join(tempDir, KEN_ALL_CSV_FILENAME);

  try {
    const tasks = createListr({
      distPath,
      pretty,
      tempDir,
      kenAllZipPath,
      kenAllCsvPath,
    });
    await tasks.run();
  } finally {
    await rm(tempDir, { force: true, recursive: true });
  }
};

interface Context {
  distPath: string;
  pretty: boolean;
  tempDir: string;
  kenAllZipPath: string;
  kenAllCsvPath: string;
  postalCodeAddressesMap?: Map<string, Address[]>;
}

function createListr(context: Context): Listr<Context> {
  return new Listr<Context>(
    [
      {
        title: 'Download "ken-all.zip".',
        async task({ kenAllZipPath }) {
          await downloadKenAllZip(kenAllZipPath);
        },
      },
      {
        title: 'Unzip "ken-all.zip',
        async task({ kenAllZipPath, tempDir }) {
          await unzip(kenAllZipPath, tempDir);
        },
      },
      {
        title: 'Create postal code addresses map',
        async task(context) {
          const { kenAllCsvPath } = context;
          const postalCodeAddressesMap = await createPostalCodeAddressesMap(
            readKenAllRecord(kenAllCsvPath)
          );
          context.postalCodeAddressesMap = postalCodeAddressesMap;
        },
      },
      {
        title: 'Write addresses json',
        async task({ postalCodeAddressesMap, distPath, pretty }, task) {
          if (postalCodeAddressesMap == null) {
            task.skip();
            return;
          }

          let completedCount = 0;
          await writeAddressesJson({
            postalCodeAddressesMap,
            distPath,
            pretty,
            onOutput() {
              ++completedCount;
              task.output = `${completedCount}/${postalCodeAddressesMap.size}`;
            },
          });
        },
      },
    ],
    {
      ctx: context,
      concurrent: false,
    }
  );
}
