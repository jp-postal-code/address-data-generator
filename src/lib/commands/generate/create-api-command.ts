import { apiAction } from '@/lib/actions/generate/api-action';
import { Argument, Command, Option } from 'commander';

export function createApiCommand() {
  const command = new Command('api');

  command
    .addArgument(new Argument('output dir', 'output directory'))
    .addOption(new Option('--pretty', 'pretty output'))
    .action(apiAction);

  return command;
}
