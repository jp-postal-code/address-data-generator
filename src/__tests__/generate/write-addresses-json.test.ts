import { writeAddressesJson } from '@/lib/generate/write-addresses-json';
import { Address } from '@/lib/types/address';
import { makeTempDir } from '@/lib/utils/make-temp-dir';
import { afterEach, beforeEach, expect, test } from '@jest/globals';
import { rmSync } from 'fs';
import { readFile } from 'fs/promises';
import { join } from 'path';

let workDir: string;

beforeEach(async () => {
  workDir = await makeTempDir();
});

afterEach(() => {
  rmSync(workDir, { force: true, recursive: true });
});

test.each([true, false])(
  'write addresses json: pretty = %s',
  async (pretty) => {
    const postalCodeAddressesMap = new Map<string, Address[]>([
      [
        '0600000',
        [
          {
            postalCode: '0600000',
            regionId: 1,
            region: '北海道',
            locality: '札幌市中央区',
            address1: '',
            address2: '',
            regionKana: 'ホッカイドウ',
            localityKana: 'サッポロシチュウオウク',
            address1Kana: '',
            address2Kana: '',
          },
        ],
      ],
      [
        '0360223',
        [
          {
            postalCode: '0360223',
            regionId: 2,
            region: '青森県',
            locality: '平川市',
            address1: '西野曽江川崎',
            address2: '',
            regionKana: 'アオモリケン',
            localityKana: 'ヒラカワシ',
            address1Kana: 'ニシノゾエカワサキ',
            address2Kana: '',
          },
          {
            postalCode: '0360223',
            regionId: 2,
            region: '青森県',
            locality: '平川市',
            address1: '西野曽江橋元',
            address2: '',
            regionKana: 'アオモリケン',
            localityKana: 'ヒラカワシ',
            address1Kana: 'ニシノゾエハシモト',
            address2Kana: '',
          },
          {
            postalCode: '0360223',
            regionId: 2,
            region: '青森県',
            locality: '平川市',
            address1: '西野曽江広田',
            address2: '',
            regionKana: 'アオモリケン',
            localityKana: 'ヒラカワシ',
            address1Kana: 'ニシノゾエヒロタ',
            address2Kana: '',
          },
        ],
      ],
    ]);

    await writeAddressesJson({
      postalCodeAddressesMap,
      distPath: workDir,
      pretty,
    });

    await expect(
      readFile(join(workDir, '060', '0000.json'), 'utf-8')
    ).resolves.toMatchSnapshot();

    await expect(
      readFile(join(workDir, '036', '0223.json'), 'utf-8')
    ).resolves.toMatchSnapshot();
  }
);
