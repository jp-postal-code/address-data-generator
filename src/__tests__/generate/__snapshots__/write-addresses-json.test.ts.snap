// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`write addresses json: pretty = false 1`] = `"[{"postalCode":"0600000","regionId":1,"region":"北海道","locality":"札幌市中央区","address1":"","address2":"","regionKana":"ホッカイドウ","localityKana":"サッポロシチュウオウク","address1Kana":"","address2Kana":""}]"`;

exports[`write addresses json: pretty = false 2`] = `"[{"postalCode":"0360223","regionId":2,"region":"青森県","locality":"平川市","address1":"西野曽江川崎","address2":"","regionKana":"アオモリケン","localityKana":"ヒラカワシ","address1Kana":"ニシノゾエカワサキ","address2Kana":""},{"postalCode":"0360223","regionId":2,"region":"青森県","locality":"平川市","address1":"西野曽江橋元","address2":"","regionKana":"アオモリケン","localityKana":"ヒラカワシ","address1Kana":"ニシノゾエハシモト","address2Kana":""},{"postalCode":"0360223","regionId":2,"region":"青森県","locality":"平川市","address1":"西野曽江広田","address2":"","regionKana":"アオモリケン","localityKana":"ヒラカワシ","address1Kana":"ニシノゾエヒロタ","address2Kana":""}]"`;

exports[`write addresses json: pretty = true 1`] = `
"[
  {
    "postalCode": "0600000",
    "regionId": 1,
    "region": "北海道",
    "locality": "札幌市中央区",
    "address1": "",
    "address2": "",
    "regionKana": "ホッカイドウ",
    "localityKana": "サッポロシチュウオウク",
    "address1Kana": "",
    "address2Kana": ""
  }
]"
`;

exports[`write addresses json: pretty = true 2`] = `
"[
  {
    "postalCode": "0360223",
    "regionId": 2,
    "region": "青森県",
    "locality": "平川市",
    "address1": "西野曽江川崎",
    "address2": "",
    "regionKana": "アオモリケン",
    "localityKana": "ヒラカワシ",
    "address1Kana": "ニシノゾエカワサキ",
    "address2Kana": ""
  },
  {
    "postalCode": "0360223",
    "regionId": 2,
    "region": "青森県",
    "locality": "平川市",
    "address1": "西野曽江橋元",
    "address2": "",
    "regionKana": "アオモリケン",
    "localityKana": "ヒラカワシ",
    "address1Kana": "ニシノゾエハシモト",
    "address2Kana": ""
  },
  {
    "postalCode": "0360223",
    "regionId": 2,
    "region": "青森県",
    "locality": "平川市",
    "address1": "西野曽江広田",
    "address2": "",
    "regionKana": "アオモリケン",
    "localityKana": "ヒラカワシ",
    "address1Kana": "ニシノゾエヒロタ",
    "address2Kana": ""
  }
]"
`;
